tokio-fs/src/file/open_options.rs

```diff
@@ -0,0 +1,209 @@
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let std = try_ready!(::blocking_io(|| {
+            self.options.open(&self.path)
+        }));
+
+        let file = File::from_std(std);
+        Ok(file.into())
+    }
+}
+
+/// Options and flags which can be used to configure how a file is opened.
+///
+/// This is a specialized version of [`std::fs::OpenOptions`] for usage from
+/// the Tokio runtime.
+///
+/// [`std::fs::OpenOptions`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html
+#[derive(Clone, Debug)]
+pub struct OpenOptions(StdOpenOptions);
+
+impl OpenOptions {
+    /// Creates a blank new set of options ready for configuration.
+    ///
+    /// All options are initially set to `false`.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// use tokio::fs::OpenOptions;
+    ///
+    /// let mut options = OpenOptions::new();
+    /// let future = options.read(true).open("foo.txt");
+    /// ```
+    pub fn new() -> OpenOptions {
+        OpenOptions(StdOpenOptions::new())
+    }
+
+    /// See the underlying [`read`] call for details.
+    ///
+    /// [`read`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.read
+    pub fn read(&mut self, read: bool) -> &mut OpenOptions {
+        self.0.read(read);
+        self
+    }
+
+    /// See the underlying [`write`] call for details.
+    ///
+    /// [`write`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.write
+    pub fn write(&mut self, write: bool) -> &mut OpenOptions {
+        self.0.write(write);
+        self
+    }
+
+    /// See the underlying [`append`] call for details.
+    ///
+    /// [`append`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.append
+    pub fn append(&mut self, append: bool) -> &mut OpenOptions {
+        self.0.append(append);
+        self
+    }
+
+    /// See the underlying [`truncate`] call for details.
+    ///
+    /// [`truncate`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.truncate
+    pub fn truncate(&mut self, truncate: bool) -> &mut OpenOptions {
+        self.0.truncate(truncate);
+        self
+    }
+
+    /// See the underlying [`create`] call for details.
+    ///
+    /// [`create`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create
+    pub fn create(&mut self, create: bool) -> &mut OpenOptions {
+        self.0.create(create);
+        self
+    }
+
+    /// See the underlying [`create_new`] call for details.
+    ///
+    /// [`create_new`]: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create_new
+    pub fn create_new(&mut self, create_new: bool) -> &mut OpenOptions {
+        self.0.create_new(create_new);
+        self
+    }
+
+    /// See the underlying [`mode`] call for details.
+    ///
+    /// [`mode`]: https://doc.rust-lang.org/beta/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode
+    #[cfg(unix)]
+    pub fn mode(&mut self, mode: u32) -> &mut OpenOptions {
```

Comment:
Instead of adding these fns, in this case, we probably could just impl `std::os::unix::fs::OpenOptionsExt`.

The logic being that this trait has no blocking function.

The counter argument would be that blocking fns could be added to the trait in `std`. However, then I think we should copy the trait and follow the extension trait pattern too, just to mirror `std`.

Thoughts?

Reply:
I did try this, but `std::os::windows::fs::OpenOptionsExt` is impossible to implement on foreign types as [one of the methods returns `std::fs::OpenOptions`][m], and didn't want to do the unix/windows stuff differently. I checked to see if there was a bug filed, which led to a [comment][c] stating the traits were "internal implementation details" and "the current goal is to replace the extension traits with inherent methods directly on the types". I figured I'd get the jump on `std` and implement it that way right now.

[m]: https://doc.rust-lang.org/1.26.2/std/os/windows/fs/trait.OpenOptionsExt.html#tymethod.security_qos_flags
[c]: https://github.com/rust-lang/rust/issues/49961#issuecomment-382751777

I didn't think of duplicating the traits, If you'd prefer to match the current `std`, rather than the anticipated future state of `std`, then I can do it that way.

Reply:
I will ask some others what they think the pattern should be here.

Reply:
The idea that came up was to leave off all the platform specific fns and traits. Instead, we can add a `From<StdOpenOptions> for OpenOptions` implementation. This way, any more advanced configuration can be done via the `std` type and then converted to the Tokio type.

What do you think of that?